Fix port forward no explicit ports

alonSadan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

- Until now, it has been possible to perform `port-forward` to
the VMI's pod, only when the target port was explicitly specified
in the VMI spec (With masquerade binding).

This commit adds nft and iptables rules to support `port-forward` also
when the port is not specified in the VMI spec,
and it will work for any port that the vmi listens to.

When the port is specified, the old behavior will be kept,
and port-forward will work only for the pre-defined port,
Making it possible for the user to limit the number of ports
that are allowed for port-forwarding.

-    Adds tests for `kubectl port-forward` to
    the VMI's pod. It includes:
    
     -  test `kubectl port-forward` to a port that is declared
     in the VMI spec.
    
     - test `kubectl port-forward` to a port that is NOT declared
     in the VMI spec.

     - test `kubectl port-forward` to a port that is NOT declared
     in the VMI spec. Should Fail because we limit the ports can be
     used by declaring them (See previous commit).


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add NFT and IPTables rules to allow port-forward to non-declared ports on the VMI. Declaring ports on VMI will limit 
the ports the can be used for port-forwarding to those that are declared.
```

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.19

alonSadan commented:
/test pull-kubevirt-e2e-k8s-1.19

alonSadan commented:
/cc @rhrazdil 
Can you take a look?

rhrazdil commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1126,8 +1126,25 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
+		}
 
-		return err
+		err = Handler.NftablesAppendRule(proto, "nat", "KUBEVIRT_POSTINBOUND",
```
in reaction to [Alona's comment](https://github.com/kubevirt/kubevirt/pull/5200#discussion_r606840280)
I updated this rule (same rule, only for VM with explicit ports) in my PR like this

```
err = Handler.NftablesAppendRule(proto, "nat", "KUBEVIRT_POSTINBOUND",
	strings.ToLower(port.Protocol),
	"dport",
	strconv.Itoa(int(port.Port)),
	Handler.GetNFTIPString(proto), "daddr", b.getVifIpByProtocol(proto),
	"counter", "snat", "to", b.getGatewayByProtocol(proto))
```

Which matches packets by destination IP (being equal to VM IP) and it eliminates danger of matching packets with external IP. After all, we want to `snat` only packets that are going into the VM, and at `postrouting` stage, these packets have the destionation IP address set to VM IP, because of the `dnat` rule in the `output` chain which happens earlier. 

cc @AlonaKaplan

alonSadan replied:
understand. Makes more sense. I changed the rule to what you suggested without the `dport` part. because the point is to not be depended on specific port.

rhrazdil replied:
> I changed the rule to what you suggested without the dport part.

and I totally missed that!

alonSadan commented:
[rebased on master](https://github.com/kubevirt/kubevirt/compare/ee2db8301255aca4e09159f1e925696aac360df2..8d4895593cec802d7565385bf341d3741811480f)

alonSadan commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1126,8 +1126,25 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
+
+		err = Handler.NftablesAppendRule(proto, "nat", "output",
+			Handler.GetNFTIPString(proto), "daddr", getLoopbackAdrress(proto),
+			"counter", "jump", "KUBEVIRT_PREINBOUND")
```
Done.

phoracek commented on tests/network/BUILD.bazel:
```diff
@@ -7,6 +7,7 @@ go_library(
         "expose.go",
         "framework.go",
         "networkpolicy.go",
+        "port-forward.go",
```
please use underscore

alonSadan replied:
Done

rhrazdil replied:
I think you renamed the file, but left the original file name in BUILD.bazel

rhrazdil commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,145 @@
+			})
+
+			It("should reach the vmi", func() {
+				vmiWithDeclaredPort, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Get(vmiWithDeclaredPort.Name, &metav1.GetOptions{})
```
Seems like `vmiWithDeclaredPort` is not used, we can remove it, right?

alonSadan replied:
oh right. Done.

rhrazdil commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,145 @@
+			})
+
+			It("should reach the vmi", func() {
+				vmiWithoutDeclaredPort, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Get(vmiWithoutDeclaredPort.Name, &metav1.GetOptions{})
```
ditto

alonSadan replied:
Done.

rhrazdil commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,145 @@
+				Expect(portForwardCmd.Start()).To(Succeed())
+			})
+
+			AfterEach(func() {
```
You could move this to the upper Context, and then you don't need it twice

alonSadan replied:
Tnx. Done.

rhrazdil commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,145 @@
+			})
+		})
+
+		When("performing port-forward from a local port to a VMI's pod non-declared port", func() {
```
Just a thought, what about using random port number from range 2000-65535?

rhrazdil replied:
Nah, it's probably nonsense, it's okay as is

alonSadan commented:
[rebased on master](https://github.com/kubevirt/kubevirt/compare/48d799b31bb4f8681000c783062a8dec6abd4de9..9edc2077eda479acd5341ceb8017b6abb5357d8e)

alonSadan reviewed (commented):
Thank you @rhrazdil . addressed your comments.

> I think you renamed the file, but left the original file name in BUILD.bazel

also fixed BUILD.bazel

alonSadan commented:
I forgot to update the unit tests after I made the changes to the nat rules. Done it [here](https://github.com/kubevirt/kubevirt/compare/5bf5580f3bc9f481a89cf8cec4c5987d38ab0651..56c566dd5e357326cda9aac5d03e4012f4e8193e).

rhrazdil commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -1123,8 +1123,25 @@ func (b *MasqueradeBindMechanism) createNatRulesUsingNftables(proto iptables.Pro
-		return err
+		err = Handler.NftablesAppendRule(proto, "nat", "KUBEVIRT_POSTINBOUND",
+			Handler.GetNFTIPString(proto), "daddr", b.getVifIpByProtocol(proto),
+			"counter", "snat", "to", b.getGatewayByProtocol(proto))
```
Hello @alonSadan , sorry for being back and forth on this. We realized one bad consequence of using the destination IP address for matching in `KUBEVIRT_POSTINBOUND`.
This rule matches traffic going into the VM from k8s pod network, so the VM sees the origin as 10.0.2.1.
Since we want he masquarading to be transparent for the VM, we don't want to change the source IP for traffic coming from the pod network.
Could you please change the rules back to what you had initially?

alonSadan replied:
Done.

alonSadan commented:
[rebased on master](https://github.com/kubevirt/kubevirt/compare/56c566dd5e357326cda9aac5d03e4012f4e8193e..c807b4beb1425949e1ac54cc1c0ba59e2b31bb5f)

alonSadan commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,133 @@
+	})
+})
+
+func newRandomCirrosVMIWithNetworkAndInterfaceWithMasqueradeBinding(ports []v1.Port) *v1.VirtualMachineInstance {
```
I thought it will be more readable this way, but I don't mind change it, it will be readable enough.
Done.

alonSadan commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,133 @@
+
+		var portForwardCmd *exec.Cmd
+		AfterEach(func() {
+			Expect(portForwardCmd.Process.Kill()).To(Succeed())
```
I understand your point, and I like what suggested to do instead. 
Changed.

alonSadan reviewed (commented):
Thanks @rhrazdil . [I addressed your comments for the second commit](https://github.com/kubevirt/kubevirt/compare/17a27ecbcd02e200a22d568a295e4e06d3f86460..d0b2eed5e59e3d45a30f78dfc24b28618681741c).

alonSadan commented:
[I increased the timeout in the tests](https://github.com/kubevirt/kubevirt/compare/d0b2eed5e59e3d45a30f78dfc24b28618681741c..c25eb0561476d98ed9af72c514f8ed5e49abb531) , because locally it passes consistently and on CI and might need more time.

alonSadan commented:
The port-forward cmd is now executed in a [smarter way](https://github.com/kubevirt/kubevirt/compare/c25eb0561476d98ed9af72c514f8ed5e49abb531..f87b21244adaf5aac0d0e88af7722186a707c9ba), chosing the correct client (oc, kubectl ) on runtime

rhrazdil commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,143 @@
+	})
+})
+
+func portForwardCommand(podName string, podNamespace string, sourcePort, targetPort int) (*exec.Cmd, error) {
```
What do you think about passing only the pod as the first parameter

alonSadan replied:
Sorry @rhrazdil  for the late reply.

I think you are right. no point of passing two strings when I can pass just the object.
changed.

alonSadan commented:
[rebased on master](https://github.com/kubevirt/kubevirt/compare/f87b21244adaf5aac0d0e88af7722186a707c9ba..6cfc76d0a97da8bad35cca53a2026f34ea2cc090)

rhrazdil commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,145 @@
+			var vmiWithoutDeclaredPort *v1.VirtualMachineInstance
+
+			BeforeEach(func() {
+				vmiWithoutDeclaredPort = createCirrosVMIWithPortsAndBlockUntilReady(virtClient, []v1.Port{})
```
This BeforeEach and the one on line 68 is practically the same.

What do you think about moving the body of this `BeforeEach` it to an upper level `JustBeforeEach`?
Then you could have this `BeforeEach` (and the one on line 68)  to only set `ports` variable accordingly - either with an explicit port or without any ports.

alonSadan replied:
ahh that's a lot nicer. Done.
What do think about it now?

rhrazdil commented:
Fix for the issue with LiveMigration was fixed, tests should pass now
/retest

alonSadan commented:
[rebased on master](https://github.com/kubevirt/kubevirt/compare/274fe42ffe66e459d6530e4cd0f7b9409c7737fd..ada4a7811ea42578aeeb3ba466d1ed550eb0cef7)

alonSadan commented:
[on this push](https://github.com/kubevirt/kubevirt/compare/ada4a7811ea42578aeeb3ba466d1ed550eb0cef7..2c8493aad53585025d2d18168e6fc25f937feb9f) I added the IPTables rules to fix the port-forward issues, in addtion to the NFTables.

In the tests I addressed @rhrazdil  [comment](https://github.com/kubevirt/kubevirt/pull/5431#discussion_r631864776) and modified the beforeEach

alonSadan commented:
Thank you Radim:)

AlonaKaplan commented:
Please updated the release note.

AlonaKaplan commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,139 @@
+		var (
+			portForwardCmd *exec.Cmd
+			vmi            *v1.VirtualMachineInstance
+			VMIport        int
```
Is there a special reason this variable is public? Anyway, I would change the name to `vmiHttpServerPort`.

AlonaKaplan commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,139 @@
+			})
+		})
+
+		When("performing port-forward from a local port to a VMI's pod non-declared port", func() {
```
Please separate this test into 2.
1. "performing port-forward from a local port to a VMI with no declared ports"
2. "performing port-forward from a local port to a VMI's non-declared port"

alonSadan replied:
Nice. Just emphasizing that on 2 it should be a negative test. The ports declaration is limiting the ports that can be used. See the previous commit message. 
I added the negative test.

AlonaKaplan commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,139 @@
+			})
+
+			It("should reach the vmi", func() {
+				defer func() {
```
Why `defer` and not `AfterEach`? Since the command was created in the `BeforeEach` I believe it makes sense to kill it in the `AfterEach`.

AlonaKaplan replied:
@rhrazdil @alonSadan I think there was a misunderstanding. The original [comment](https://github.com/kubevirt/kubevirt/pull/5200#discussion_r617235464) explained why we are using `tests.BeforeTestCleanup` and not cleaning all the entities on the `AfterEach` of the tests. The main reason is that the reports are collected after the teardown.
This is the standard. And if we have a new type of entity that is not cleaned by `tests.BeforeTestCleanup` we prefer to add it to `tests.BeforeTestCleanup` instead of doing a specific cleaning in the `AfterEach`. However, here a process is cleaned, not an entity (am I right? is there any entity created while this process is running?). I don't think we collect information about the running processes and even if we do, did you verify using `defer` will help the situation and the report is collected after it (I believe that now the cleanup we be executed even vefore the `AfterEach`)?

alonSadan replied:
You are right. There is no cleanup in the defer so i is redundant I think.
Thanks.
Moving back to afterEach.

rhrazdil replied:
Ah I see, thanks for the explanation @AlonaKaplan , I indeed misunderstood the the origianal comment.
Sorry for misleading you @alonSadan!

alonSadan replied:
all good. I understood it the same way you did.

AlonaKaplan commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,139 @@
+		return nil
+	}
+
+	return portForwardCmd.Process.Kill()
```
I see that [here](https://github.com/kubevirt/kubevirt/blob/fd39d9e22a57c6391e03fd8c496e98560d23490f/tests/storage/imageupload.go#L179) there is also a `wait` for the command to be terminated. What is the reason for not having it here as well?
According to the [documentation](https://golang.org/pkg/os/#Process.Kill) - "kill does not wait until the Process has actually exited". So if we want to be sure the port-forwarding is kill before the next command is invoked, I believe we should add a `portForwardCmd.Wait`.

alonSadan replied:
To be on the safe side modified the test to use different local ports. That's the only dangerous thing I could think of, but still I don't think we need to wait.
Is that acceptable?

AlonaKaplan replied:
Why not using `wait`? IMO, maintaining there is no overlap between tests can be difficult/dangerous. Using `wait` is much simpler.

alonSadan replied:
I thought it will be faster, but I guees it is not that significant. Will change to `wait()`

alonSadan replied:
[Done](https://github.com/kubevirt/kubevirt/compare/d65db08009e6f838e208f0312576f23305371196..3e5d913addf7506b3c2738401bb7a99677ba1035)

AlonaKaplan commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,139 @@
+			portForwardCmd *exec.Cmd
+			vmi            *v1.VirtualMachineInstance
+			VMIport        int
+			ports          []v1.Port
```
I would change to `vmiDeclatedPorts`.

AlonaKaplan commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,139 @@
+
+				By(fmt.Sprintf("checking that service running on port %d can be reached", nonDeclaredPort))
+				Eventually(func() error {
+					err = exec.Command("curl", fmt.Sprintf("127.0.0.1:%d", localPort)).Run()
```
Can be exported to a method  `testConnectivity` (or something like this), since both of the tests have it.

alonSadan commented:
[rebased on master](https://github.com/kubevirt/kubevirt/compare/2c8493aad53585025d2d18168e6fc25f937feb9f..fa27590b2bb5e23172476b4062627739b2aba476)

alonSadan commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,139 @@
+				}()
+
+				By(fmt.Sprintf("checking that service running on port %d can be reached", declaredPort))
+				Eventually(func() error {
```
I think the VM can be considered ready (by `createCirrosVMIWithPortsAndBlockUntilReady()`) even though it is still haven't finish It's booting process. The VM is considered ready if there was a successfull login (see [here](https://github.com/kubevirt/kubevirt/blob/master/tests/utils.go#L3064)). Login to the VMI can be done even when the VMI is still booting up,  So the HTTP server might not have started when we reach the curl command. So I thought to use eventually to wait for that. 
In my local env the tests usually won't pass without `Eventually` and constantly passing with `Eventually`  and I think it is because of that.

AlonaKaplan replied:
You start the http server in the `BeforeEach` so I don't understand how the server can still be down in the stage of the `curl`.
We use `WaitUntilVMIReady` in most of our tests and we don't need eventually there. For example - [this](https://github.com/kubevirt/kubevirt/blob/a532f588a293d0aee91c22161189214f2b6c1850/tests/network/vmi_networking.go#L767).
I believe the reason you need eventually is different. Please debug it.

alonSadan replied:
You are right. The reason I gave does not holds because we couldn't start the http server without the vmi to be ready and finish boot. 
It could be also that the port-forward command takes time to work. But Since I can not wait for it to finish and return a OK status, and I have to wait somewhere i think waiting in the Eventually won't be so bad. 

Even In the example you shared there are actually a lot of timeouts as part of the `verifyClientServerConnectivity()` func and the fact that it is using expecters which uses timeouts.

The only suspicious thing I thought of was that the port-forward command from a previous test still ha havn't finished (like you noted [here](https://github.com/kubevirt/kubevirt/pull/5431/commits/de977707f5a5071458ed10a0e7a747cd116a0da6#r636101749) but this is not true as without eventually it fails also when I run only one test.

What do you think?

AlonaKaplan replied:
> It could be also that the port-forward command takes time to work
Let's not depend on assumption. I think we should understand why the first tries are not working.

> Even In the example you shared there are actually a lot of timeouts
There is a difference between a timeout and an eventually.
Timeout means the operation may take time, but failures are not accepted.
`Eventually` means the operation may fail several time before it succeeds.

Using `Eventually` may hide bugs. IMO, it's ok to use it only if we understand the exact reason why it is needed.

alonSadan replied:
Understand.
Thanks for explaining.

still the command takes sometime to act, and in order for the test to pass we need to wait a bit for it, before we try the curl.
I removed the `Eventually` and added 3 seconds sleep before the curl. it should be enough. 
that way we will still wait but not ignore failures.
What do you think?

alonSadan replied:
Removed the `Eventually` and now waiting for the port-forward cmd output.
https://github.com/kubevirt/kubevirt/compare/699f4e38bbaa7be7bf79da628385590f98f1faa4..2d512ffdf7688e2cd00402f51d783d38e9ce7913

alonSadan reviewed (commented):
Thanks @AlonaKaplan . [Addressed some of the comments](https://github.com/kubevirt/kubevirt/compare/fa27590b2bb5e23172476b4062627739b2aba476..de977707f5a5071458ed10a0e7a747cd116a0da6) and responded to some.

alonSadan commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,139 @@
+				}()
+
+				By(fmt.Sprintf("checking that service running on port %d can be reached", nonDeclaredPort))
+				Eventually(func() error {
```
same as https://github.com/kubevirt/kubevirt/pull/5431/commits/de977707f5a5071458ed10a0e7a747cd116a0da6#r637777939

alonSadan reviewed (commented):
[Made each test to use different local ports and changed the `defer` to `afterEach`.](https://github.com/kubevirt/kubevirt/compare/de977707f5a5071458ed10a0e7a747cd116a0da6..39f580554b84fd5ccd683741e94e1a10c8d65673)
Also replied to @AlonaKaplan comments.

alonSadan commented:
[tests commit: made each test wait for the port-forward process to finish after killing it before moving to the next test](https://github.com/kubevirt/kubevirt/compare/d65db08009e6f838e208f0312576f23305371196..3e5d913addf7506b3c2738401bb7a99677ba1035)

alonSadan commented:
[Removed the `Eventually` and instead added a 3 seconds sleep before we test the port-forward cmd, since it takes some time to act.](https://github.com/kubevirt/kubevirt/compare/3e5d913addf7506b3c2738401bb7a99677ba1035..742535c070c84e9e8c961472a1cc3cfa60eb6aec)

@AlonaKaplan is that OK?

AlonaKaplan commented:
> [Removed the `Eventually` and instead added a 3 seconds sleep before we test the port-forward cmd, since it takes some time to act.](https://github.com/kubevirt/kubevirt/compare/3e5d913addf7506b3c2738401bb7a99677ba1035..742535c070c84e9e8c961472a1cc3cfa60eb6aec)
> 
> @AlonaKaplan is that OK?

Isn't there a better way to find out `port-forward` is active?
Maybe tracking the output already has something like `Forwarding from 127.0.0.1:28015 -> 27017`?

alonSadan commented:
> > [Removed the `Eventually` and instead added a 3 seconds sleep before we test the port-forward cmd, since it takes some time to act.](https://github.com/kubevirt/kubevirt/compare/3e5d913addf7506b3c2738401bb7a99677ba1035..742535c070c84e9e8c961472a1cc3cfa60eb6aec)
> > @AlonaKaplan is that OK?
> 
> Isn't there a better way to find out `port-forward` is active?
> Maybe tracking the output already has something like `Forwarding from 127.0.0.1:28015 -> 27017`?
> 
> EDIT: you can see and example [here](https://github.com/kubevirt/kubevirt/blob/3a059eccec70e9e2b11c14a9fa963eb356901306/pkg/virt-launcher/virtwrap/util/libvirt_helper.go#L214) how to read from the `StrerrPipe` of a command (there is also similar `StdoutPipe`).

Nice!!
Thanks Alona.
[Done.](https://github.com/kubevirt/kubevirt/compare/699f4e38bbaa7be7bf79da628385590f98f1faa4..2d512ffdf7688e2cd00402f51d783d38e9ce7913)

alonSadan commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,167 @@
+
+			It("should not reach the vmi", func() {
+				By(fmt.Sprintf("checking that service running on port %d can not be reached", nonDeclaredPort))
+				Consistently(func() error {
```
is was required before I waited for the port-forward cmd. Now it is redundant and forgot to remove it.
Removed.

alonSadan commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,167 @@
+}
+
+func testConnectivityThroughLocalPort(portNumber int) error {
+
```
Done.

alonSadan commented:
[removed consistently from negative test and redundant empty line. ](https://github.com/kubevirt/kubevirt/compare/2d512ffdf7688e2cd00402f51d783d38e9ce7913..89e967b639663619cc895c1b5840c7e81ade3492)

AlonaKaplan commented on tests/network/port_forward.go:
```diff
@@ -0,0 +1,163 @@
+
+			It("should not reach the vmi", func() {
+				By(fmt.Sprintf("checking that service running on port %d can not be reached", nonDeclaredPort))
+				Expect(testConnectivityThroughLocalPort(localPort)).To(Succeed())
```
Shouldn't it fail?

alonSadan replied:
My mistake.
Fixed.

alonSadan commented:
[rebased on master](https://github.com/kubevirt/kubevirt/compare/89e967b639663619cc895c1b5840c7e81ade3492..b99cd77a4d980a8bbf176dfb5850c04689bd6615)

[fixed a mistake I made by asserting the negative test to pass](https://github.com/kubevirt/kubevirt/compare/b99cd77a4d980a8bbf176dfb5850c04689bd6615..5ce7422738bdfb9af7e9de2c703596c89ed629eb)

AlonaKaplan commented:
/approve

alonSadan commented:
[rebased on  master](https://github.com/kubevirt/kubevirt/compare/5ce7422738bdfb9af7e9de2c703596c89ed629eb..c1fed48158390dfa83961b5ab570ad7cfd3443ca)

rhrazdil reviewed (approved):
Nice work
/lgtm
